File: lib/tree_diameter.cpp
#include <algorithm>
#include <climits>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <vector>
using namespace std;

using ll = long long;

struct Edge {
  ll to;
  ll cost;
};
using Graph = vector<vector<Edge>>;

// pair<最遠点間距離, 最遠点> を求める
pair<ll, ll> dfs(Graph &G, ll idx, ll parent) {
  pair<ll, ll> ret(0, idx);
  for (auto &e : G[idx]) {
    if (e.to == parent) {
      continue;
    }
    auto cost = dfs(G, e.to, idx);
    cost.first += e.cost;
    if (ret.first < cost.first) {
      ret = cost;
    }
  }
  return ret;
}

// 重み付き木の直径を求める.
// (重みが与えられていない場合は, 全ての重みを1とする.)
// 頂点数をN, O(N).
ll tree_diameter(Graph &G) {
  auto p = dfs(G, 0, -1);
  auto q = dfs(G, p.second, -1);
  return q.first;
}

// http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=GRL_5_A&lang=jp
int main() {
  // ios::sync_with_stdio(false);
  // cin.tie(0);
  ll N;
  cin >> N;
  Graph G(N);

  for (ll i = 0; i < N - 1; i++) {
    ll s, t, w;
    cin >> s >> t >> w;
    G[s].push_back(Edge{t, w});
    G[t].push_back(Edge{s, w});
  }

  ll ans = tree_diameter(G);
  cout << ans << endl;
  return 0;
}

File: python/lib/tree_diameter.py
from typing import List, Tuple
import sys

sys.setrecursionlimit(10**7)

INF = float("inf")


class Edge:
    def __init__(self, to: int, cost: int) -> None:
        self.to = to
        self.cost = cost


def dfs(G: List[List[Edge]], idx: int, parent: int) -> Tuple[int, int]:
    ret = (0, idx)
    for e in G[idx]:
        if e.to == parent:
            continue
        cost, far = dfs(G, e.to, idx)
        cost += e.cost
        if ret[0] < cost:
            ret = (cost, far)
    return ret


def tree_diameter(G: List[List[Edge]]) -> int:
    p = dfs(G, 0, -1)
    q = dfs(G, p[1], -1)
    return q[0]


def main() -> None:
    N = int(input())
    G: List[List[Edge]] = [[] for _ in range(N)]

    for _ in range(N - 1):
        s, t, w = map(int, input().split())
        G[s].append(Edge(t, w))
        G[t].append(Edge(s, w))

    ans = tree_diameter(G)
    print(ans)


if __name__ == "__main__":
    main()
